### PR TITLE
fix(resolver): reclassify status to error when all patch tools failed despite dirty worktree

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Problem

The gptme resolver currently trusts the model's `RESOLVER_STATUS: changes` marker whenever the worktree is dirty — even when `gptme-stdout.log` shows every `patch`/ `save` tool call failed with an execution error. This causes the workflow to open a bogus draft PR containing only a previous commit's stale changes (e.g. a submodule bump), wasting reviewer time.

## Root Cause

The canary #824 artifact analysis (ErikBjare/bob#824) showed the model emitted `RESOLVER_STATUS: changes` while both `patch` calls returned `Error during execution: Patch failed: original chunk not found in file`. The worktree was dirty only because gptme's `--no-confirm` checked out and bumped a submodule pin before any meaningful edit. The dirty-worktree check alone is insufficient.

## Solution

**L2(b) — orchestrator-level tool-error detection**: before accepting `RESOLVER_STATUS: changes` + dirty worktree, scan `gptme-stdout.log` for `Error during execution` lines. If ALL tool calls errored, downgrade to `error` with the errors surfaced in the message. If at least one `patch`/ `save` succeeded, accept the result as-is.

A new `extract_tool_errors()` helper parses the gptme log for 'Error during execution' markers, deduplicates consecutive identical errors, and returns them as a structured summary. The main flow inserts a check after the existing dirty-worktree gate: if gptme said 'changes', the worktree is dirty, but every tool call errored, the status drops to 'error' with the full error summary as the message.

Also strengthens the prompt template: the 'Validate your changes took effect' section now explicitly warns against the inline-stdout hallucination pattern documented in ErikBjare/bob session b7fd.

## Verification

- 22 resolver tests pass (4 new: `test_extract_tool_errors_*`)
- No existing behavior changed — the new check only triggers when ALL tool calls errored
- Deterministic test: a gptme output with two consecutive 'Patch failed' errors correctly extracts and deduplicates them

### CODE REVIEW

File: scripts/github_resolver/resolve_issue.py [hunk 1]
```diff
@@ -509,10 +536,22 @@ def main(argv: list[str] | None = None) -> int:
...
-        message = (
-            "gptme reported changes but no files were modified. "
-            f"Upstream message: {message}"
-        )
+        # Surface tool execution errors so the issue comment shows actionable
+        # diagnostics without downloading the session artifact.  If no tool
+        # errors are found, fall back to the generic description.
+        tool_errors = extract_tool_errors(gptme_output or "")
+        if tool_errors:
+            errors_block = "\n".join(f"- `{e}`" for e in tool_errors)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_resolver/resolve_issue.py:632]: Fixed in 3405d4e. The fix is now wired to the correct branch.

Added `count_write_tool_calls()` that counts ```patch`/```save` tool invocations in gptme's stdout log. In the `worktree_dirty` success path, we now check: if `write_calls > 0 and len(errors) >= write_calls`, all writes failed — downgrade to error and fall through to the partial-work path instead of opening a bogus draft PR.

The "all errors" guard is implemented: the condition requires `len(errors) >= write_calls` (not just "any error"), so a partial-failure session where some patches succeeded still opens the PR normally.

File: scripts/github_resolver/test_resolve_issue.py [hunk 2]
```diff
@@ -157,6 +157,38 @@ def test_status_regex_ignores_prose_mentions():
...
+    assert len(errors) == 1
+
+
+def test_extract_tool_errors_handles_empty_string():
+    assert resolve_issue.extract_tool_errors("") == []
+
+
 def test_open_draft_pr_retrigger_returns_existing_url(monkeypatch):
     # On re-trigger, `gh pr create` exits 1 ("a pull request … already exists").
     # open_draft_pr must catch CalledProcessError and return the existing PR URL
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_resolver/test_resolve_issue.py:407]: Fixed in 3405d4e. Added `test_main_reclassifies_status_when_all_writes_errored_despite_dirty_worktree` — an integration test that covers the exact canary-#824 scenario end-to-end:

- Sets up a local bare repo + clone
- `fake_run_gptme` creates an incidental file (simulating a submodule bump) and returns stdout with a patch tool call that errored
- Asserts that `open_draft_pr` is never called
- Asserts the issue comment contains the error details

Also added unit tests for `count_write_tool_calls`: zero when no write blocks present, counts both `patch` and `save`, handles empty string.

**Comment by TimeToBuildBob** [comment]
Fixed in f859ee5.

The deduplication bug is now resolved:

- Added `count_tool_errors()` that counts **raw** (non-deduplicated) error lines from the gptme log — so 2 identical `"Patch failed"` errors count as 2, not 1.
- The guard in the `worktree_dirty` path now uses `error_count >= write_calls` (raw count) instead of `len(tool_errors) >= write_calls` (deduplicated count). `extract_tool_errors()` still returns the deduplicated list for readable error display.
- Added `test_count_tool_errors_counts_repeated_identical_errors` unit test confirming the raw count is 2 for two identical errors.
- Added `test_main_reclassifies_status_when_two_identical_write_errors_despite_dirty_worktree` — an integration test covering the exact canary-#824 scenario: 2 write calls, 2 identical "Patch failed" errors → guard fires → no bogus draft PR opened.
- Extracted shared git repo setup into a `local_git_repo` pytest fixture + `_run_resolver` helper to eliminate the test duplication jscpd was flagging.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: scripts/github_resolver/resolve_issue.py [hunk 3]
```diff
@@ -178,6 +178,61 @@ def parse_status(gptme_output: str) -> tuple[str, str]:
...
+def count_tool_errors(gptme_output: str) -> int:
+    """Count raw (possibly repeated) tool execution errors in gptme's stdout log.
+
+    Unlike :func:`extract_tool_errors`, this does NOT deduplicate — two identical
+    ``Error during execution`` lines count as two errors.  This is the right
+    operand to compare against :func:`count_write_tool_calls` to decide whether
+    ALL write calls failed, because gptme emits one error line per failing tool
+    invocation even when the error message is identical.
+    """
+    return len(_ERROR_RE.findall(gptme_output))
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_resolver/resolve_issue.py]: Fixed in 0479a1a. `count_tool_errors` now uses `_WRITE_TOOL_ERROR_RE` which matches only `"Patch failed|save failed"` patterns — shell commands, bash errors, and read-tool failures are excluded. Added `test_count_tool_errors_ignores_non_write_tool_errors` to guard the false-positive path (shell error alongside a successful patch should NOT trigger the downgrade).

File: scripts/github_resolver/resolve_issue.py [hunk 4]
```diff
@@ -482,37 +537,67 @@ def main(argv: list[str] | None = None) -> int:
...
+    # at least one write tool call succeeded.  If every patch/save errored, the
+    # worktree is dirty for incidental reasons (e.g. a submodule bump) and
+    # opening a draft PR would be bogus.
     if status == "changes" and worktree_dirty:
-        if args.dry_run:
-            print(f"[dry-run] Would push {branch} and open draft PR: {message}")
+        tool_errors = extract_tool_errors(gptme_output or "")
+        write_calls = count_write_tool_calls(gptme_output or "")
+        error_count = count_tool_errors(gptme_output or "")
+        if write_calls > 0 and error_count >= write_calls:
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_resolver/resolve_issue.py:590]: Fixed in 0479a1a alongside the deduplication fix. `count_tool_errors` now uses `_WRITE_TOOL_ERROR_RE = re.compile(r"Error during execution:\s*(?:Patch failed|save failed)", ...)` — a session with 1 successful patch + 1 unrelated bash error yields `error_count=0`, `write_calls=1`, condition `0 >= 1` is False, so the PR opens normally. No false positive.

**Comment by TimeToBuildBob** [comment]
@greptileai review
